File: legacy/bff/src/services/minicash.ts
import axios from "axios";
import url from "url";
import { config } from "../../config";
import { CustomLogger } from "./logger";
import { NotFoundError } from "../domain/errors/not-found-error";
import { BadRequestError } from "../domain/errors/bad-request-error";
import { UUID } from "crypto";
import { findConsumerById } from "./consumer";

const client = axios.create({
  baseURL: config.minicashBaseURL,
  timeout: 50000,
});
const logger = new CustomLogger("minicash");

export const getConsumerMinicashData = async (national_id: string) => {
  const params = new url.URLSearchParams({
    username: config.minicashUsername,
    password: config.minicashPassword,
    grant_type: "password",
  });
  logger.debug("get consumer data minicash");

  let consumer = null;
  try {
    const bearer_token = (
      await client.post(config.minicashLoginURL, params.toString(), {
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
      })
    ).data.access_token;

    logger.debug("got token");

    consumer = (
      await client.get(config.scoringEngineGetUserDataURL, {
        params: {
          ssn: national_id,
        },
        headers: {
          Authorization: `Bearer ${bearer_token}`,
        },
      })
    ).data.Data;

    logger.debug(`got consumer with ssn:  ${consumer.Ssn} and mobile: ${consumer.MobileNo}`);
  } catch (error) {
    console.error(error);
    logger.error(`Failed to get user minicash data with error ${error}`);
  }
  return consumer;
};

export const getLoansFromMiniCash = async (consumerId: UUID) => {
  try {
    logger.debug("getting consumer loans for consumer with ID: " + consumerId);
    const { national_id } = await findConsumerById(consumerId);
    const res = await client.post("/PaymentGateWay/api/MyloPayment/GetOrdersBySSN", {
      SSN: national_id,
    });
    return res.data;
  } catch (e) {
    if (e.response?.status === 404) {
      throw new NotFoundError(`Loans do not exist for SSN this consumer`);
    } else if (e.response?.status === 400) {
      console.log("bad request happened");
      throw new BadRequestError(e.data?.ErrMsg || "Bad request");
    }
    throw e;
  }
};

File: legacy/service-department/src/services/mini_cash.py
import requests
from config.settings import settings
from src.services.logging import logger

mini_cash_base_url = settings.get("mini_cash", default={}).get("base_url", default="")
mini_cash_username = settings.get("mini_cash", default={}).get("username", default="")
mini_cash_password = settings.get("mini_cash", default={}).get("password", default="")


logger = logger.bind(service="minicash", context="setup", action="setup")


class MiniCash:

    @staticmethod
    def __login() -> str:
        logger.debug("login to mini cash")
        try:
            url = mini_cash_base_url + "/Token"
            headers = {"Content-Type": "application/x-www-form-urlencoded"}
            payload = {"username": mini_cash_username, "password": mini_cash_password, "grant_type": "password"}
            response = requests.post(url, data=payload, headers=headers, timeout=30)
            if response.status_code == 200:
                logger.debug("logged in to mini cash")
                return response.json()["access_token"]
            logger.error(
                f"failed to login to mini cash with url {url}, status code {response.status_code}, payload {payload}, and response text {response.text}"
            )
            return ""
        except Exception as expt:
            logger.error("An error occurred during minicash login: %s" % str(expt))
            return ""

    @staticmethod
    def migrateConsumer(ssn: str) -> bool:
        logger.debug("get consumer credit limit")
        mini_cash_token = MiniCash.__login()
        if mini_cash_token == "":
            return False
        try:
            url = mini_cash_base_url + "/api/Client/migrate-consumer"
            headers = {"Authorization": "Bearer " + mini_cash_token, "Content-Type": "application/json"}
            payload = {"SSN": ssn}
            response = requests.post(url, json=payload, headers=headers, timeout=180)
            if response.status_code == 200:
                return True
            else:
                logger.error("failed to migrate consumer with ssn %s" % ssn)
                logger.error("response status code %s" % response.status_code)
                logger.error("response text %s" % response.json()["Message"])
                return False

        except Exception as expt:
            logger.error("An error occurred during consumer migration: %s" % str(expt))
            return False
